CIP-????/README.md

```diff
@@ -0,0 +1,471 @@
+  - Force the scrutinee (type instantiation)
+  - Evaluate the resulting function, evaluate the argument, apply the argument (x2 for two branch arguments)
+  - Evaluate the branch function, evaluate the argument, apply the argument (x2 for two constructor arguments)
+  - Enter the branch body
+- P4
+  - Evaluate the scrutinee
+  - Look at the tag
+  - Evaluate the case branch (x2 for two branch arguments)
+  - Apply the branch function to the constructor arguments (x2 for two constructor arguments)
+  - Enter the branch body
+
+Scott encoding isn't doing anything clearly unnecessary: it's quite efficient at constructing values (because it just returns a function closure right away), and it's quite efficient at deconstructing values (because it just loads the constructor arguments into an environment and then applies the branch).
+
+In particular, both ways we have to do a similar amount of work in a) evaluating the branch function, b) applying it to its arguments.
+
+#### Why is sums-of-products better at all?
+
+There are a few advantages for sums-of-products.
+The most important is that it does not evaluate all the case branches, only the one it needs. Whereas Scott-encoding passes the case branches as simple function arguments, so they are always evaluated before we proceed.
+
+This makes a surprising amount of difference.
+We are performing so few steps already for each datatype match that adding a few more to evaluate unused case branches makes a difference.
+
+#### What about binding case variables?
+
+Binding case variables allows us to get rid of some of the work identified in the previous sections.
+A `constr` that binds case variables has two differences:
+1. The body of the case branch is _known_, rather than potentially being an arbitrary lambda that we will need to resolve by doing evaluation.
+2. The constructor arguments can be loaded into the environment _all in one go_, rather than taking requiring multiple steps through the evaluator for each evaluation.
+
+In practice it seems that difference 1 makes a significant amount of difference, because even if the case branch is a literal lambda, we are forced to allocate a value for the lambda.
+Optimizing the evaluator to avoid this allocation removes a significant part of the advantage.
+
+Difference 2 does not seem to make as large a difference. If we did see a big difference here then we might want to investigate adding multi-lambdas to Plutus Core in order to gain this benefit in other places. In practice, however, it does not seem to be that significant, and prototypes of multi-lambdas have not performed well.
+
+### Appendix 3: Benchmark results
+
+#### Nofib
+
+These benchmarks are re-compiled from Haskell each time, so the comparison does not represent faster evaluation of the same script, but rather than we can now compile datatype operations using the new terms, which are faster overall.
+
+| Script             | `master` | `sums-of-products` | Change |
```

Comment:
Is there a link to the place where this table was produced? There have been so many versions of sum-of/and-products, especially when it comes to implementation, that it would make sense to reference the particular implementation.

Reply:
That would be good, but TBH my branches are a bit of a mess... I should pin a particular branch and link to it from here.

Reply:
Incorporated as a TODO